OSPSuite.Presentation.Importer/Services/IImporter.cs

```diff
@@ -0,0 +1,107 @@
+   {
+      private readonly IoC _container;
+      private readonly IDataSourceFileParser _parser;
+
+      public Importer( IoC container, IDataSourceFileParser parser)
+      {
+         _container = container;
+         _parser = parser;
+      }
+
+      public IEnumerable<IDataFormat> AvailableFormats(IUnformattedData data, IReadOnlyList<ColumnInfo> columnInfos)
+      {
+         return _container.ResolveAll<IDataFormat>()
+            .Where(x => x.SetParameters(data, columnInfos));
+      }
+
+      public void AddFromFile(IDataFormat format, IReadOnlyDictionary<string, IDataSheet> dataSheets, IReadOnlyList<ColumnInfo> columnInfos, IDataSource alreadyExisting)
+      {
+         var dataSets =
+            dataSheets
+               .ToDictionary
+               (
+                  s => s.Key,
+                  s =>
+                     new DataSet()
+                     {
+                        Data = format.Parse(s.Value.RawData, columnInfos)
+                     } as IDataSet
+               );
+         foreach (var element in dataSets)
+         {
+            var current = alreadyExisting.DataSets.GetOrAdd(element.Key, _ => new DataSet());
+            current.Data = current.Data.Union(element.Value.Data).ToDictionary(s => s.Key, s => s.Value);
+         }
+      }
+
+      public IDataSourceFile LoadFile(IReadOnlyList<ColumnInfo> columnInfos, string fileName)
+      {
+         //var filename = _dialogCreator.AskForFileToOpen(Captions.Importer.PleaseSelectDataFile, Captions.Importer.ImportFileFilter, Constants.DirectoryKey.OBSERVED_DATA, fileName);
+         //in the presenter : if string == "" (Cancel clicked), then do not try to parse
+
+
+         var dataSource = _parser.For(fileName);
+         dataSource.AvailableFormats = AvailableFormats(dataSource.DataSheets.ElementAt(0).Value.RawData, columnInfos).ToList();
+         dataSource.Format = dataSource.AvailableFormats.FirstOrDefault();
+         //TODO: check that all sheets are supporting the formats...
+         
+         return dataSource;
+      }
+
+      public IEnumerable<string> NamesFromConvention
+      (
+         string namingConvention,
+         string fileName,
+         IDictionary<string, IDataSet> dataSets,
+         IEnumerable<MetaDataMappingConverter> mappings
+      )
+      {
+         fileName = Path.GetFileNameWithoutExtension(fileName);
+         var counters = new Dictionary<string, int>();
+         return dataSets.SelectMany(ds => ds.Value.Data.Select(s =>
+         {
+            var key = mappings.Aggregate
+            (
+               new MetaDataMappingConverter()
+               {
+                  Id = namingConvention.Replace("{File}", fileName).Replace("{Sheet}", ds.Key)
+               },
+               (acc, x) =>
+                  new MetaDataMappingConverter()
+                  {
+                     Id = acc.Id.Replace($"{{{x.Id}}}", $"{s.Key.FirstOrDefault(md => md.Id == x.Index(ds.Key))?.Value}")
```

Comment:
I don't undertstand what this code is doing here

Reply:
I have added some comments but I believe the whole code will be much more readable once we refactor the Parse function